### PR TITLE
Capture fsserver errors

### PR DESCRIPTION
**Pull request by wehimwich**
This is just a test of the concept. It seems to work. I have not tried to hunt down edge cases, yet. I am actually not sure where to look for those. Suggestions would be welcome. No doubt it needs some clean-up, if not complete re-writing. Since I don't completely understand the server, I may have gone off on a weird trajectory.

I tried to make it so that the file used to record the errors would be deleted if there was a clean termination (apparently that includes `SIGINT`). Whether I selected the best place to do that is not clear to me.

One thing it does not do is capture the exit code. I suppose that might be useful if, for example it gets killed by an unhandled signal. @dehorsley suggested that it might be a good idea to run the server as a service. Maybe that would capture the exit code, as well as other output and maybe without all the file descriptor machinations. If that would be _systemd_, it might be problematic for _Wheezy_ and _Lenny_ (we might not even want to think about _Etch_) since they do not use _systemd_, at least not by default. So a service might not be a universal solution in the short-term.

### CODE REVIEW

**Review by dehorsley — COMMENTED** [review]
Looks pretty good. I'll push some changes in a minute to clean up some style

File: fsserver/server.c [hunk 1]
```diff
@@ -1088,6 +1090,10 @@ int server_new(server_t **new) {
 }
 
 void server_shutdown(server_t *s) {
+
+	if (unlink(fsserver_err_file))
+		fatal("unlinking fsserver.err file", strerror(errno));
+
```
[the remark this answers is not in the document]
Comment (dehorsley) [reply fsserver/server.c]: I decided this needs to go in `server_destroy`, in case there are any errors during the shutdown process.

**Comment by wehimwich** [comment]
Nicely done. Minor comments to follow.

File: fsserver/fsserver.c [hunk 2]
```diff
@@ -42,7 +45,11 @@
 #include "server.h"
 
 static bool opt_daemon = true;
-int error_fd           = -1;
+
+// error_fd is set to write end of pipe in the daemon process, which can be
+// used to send an message to the calling process if an error occurs early in
+// the startup of the daemon.
+int error_fd = -1;
```
Comment (wehimwich) [inline fsserver/fsserver.c:52]: maybe `static int error_fd = -1;`?

File: fsserver/fsserver.c [hunk 3]
```diff
@@ -291,6 +297,57 @@ void install_shims() {
...
+	}
+
+	if (dup2(fd_err, STDOUT_FILENO) < 0 || dup2(fd_err, STDERR_FILENO) < 0)
+		fatal("connecting to fsserver.err file", strerror(errno));
+}
+
+void setup_foreground_log(char *path) {
+	char *linep = NULL;
+	if (asprintf(&linep, "/usr/bin/tee %s", path) < 0)
+		fatal("making tee command", "asprintf");
```
Comment (wehimwich) [inline fsserver/fsserver.c:333]: You changed the other instance of this error check to use `strerror(errno)` in the `fatal()` call. Was not changing it here on purpose?

I had not used `strerror(errno)` because the _man_ page for `asprintf()` does not mention setting `errno` on an error whereas typically other _man_ pages do. Perhaps that is just a "typo."

**Comment by wehimwich** [comment]
Limited testing seemed good: both foreground and background and both with an error and without.

File: fsserver/fsserver.c (hunk 3 above)
Comment (dehorsley) [reply fsserver/fsserver.c:333]: Good point. Looking into the code, it does set errno (via the other libc functions it call), so it should give something sensible.

File: fsserver/fsserver.c (hunk 2 above)
Comment (dehorsley) [reply fsserver/fsserver.c:52]: Ah yes indeed.
